x-pack/plugins/enterprise_search/public/applications/shared/use_local_storage/use_local_storage.ts

```diff
@@ -0,0 +1,48 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { useState, useEffect } from 'react';
+
+export const useLocalStorage = <T>(key: string, defaultValue: T) => {
+  const [item, setItem] = useState<T>(getFromStorage());
+
+  function getFromStorage() {
+    const storedItem = window.localStorage.getItem(key);
+    let toStore: T = defaultValue;
+
+    if (storedItem !== null) {
+      try {
+        toStore = JSON.parse(storedItem) as T;
+      } catch (err) {
+        window.localStorage.removeItem(key);
+      }
+    }
+
+    return toStore;
+  }
+
+  const updateFromStorage = () => {
+    const storedItem = getFromStorage();
+    setItem(storedItem);
+  };
+
+  const saveToStorage = (value: T) => {
+    window.localStorage.setItem(key, JSON.stringify(value));
+    updateFromStorage();
+  };
+
+  useEffect(() => {
+    const handler = (event: StorageEvent) => {
+      if (event.key === key) {
+        updateFromStorage();
+      }
+    };
+    window.addEventListener('storage', handler);
+    return () => window.removeEventListener('storage', handler);
+  }, []);
```

Comment:
This is probably because I'm very tired today but I'm failing to understand why this helper has to be a React hook specifically instead of just being a generic vanilla JS (either a set of reusable fn's, or a class if state is really needed for some reason).

In this particular use case (the documents customization modal), I feel like the localStorage events are super straightforward:

- On page load, get the keys from storage and populate fields
- On modal save, set the updated keys to storage

I don't understand why we particularly need a listener - the use case it seems to apply to (based on my guessing + the tests) is if someone has two tabs/windows open and updates tab 1, apply the changes to tab 2. My concerns here are whether that's really something we really need from a UX perspective or if it's overengineering, especially considering the baseline of the rest of our app in terms of reactivity/dynamic-ness between tabs (or even between refreshes).

That being said, I know the code+tests are already written and it already works well, so disregard this entire comment if you like :) I think we could have gotten away just fine with a very basic set of `window.localStorage.x` function helpers that JSON.parse/stringify for us, but definitely appreciate the extra mile.

Reply:
> I'm failing to understand why this helper has to be a React hook

It's a React hook because it was a React hook when I migrated it. I am not looking to rewrite things, I'm looking to simply migrate them from one project to another. I saw a low hanging fruit, which was the pre-existing `useLocalStorage` helper that matched exactly the API of what we were using before so I simply borrowed from it. I think that is a very reasonable approach.

>  I think we could have gotten away just fine with a very basic set of window.localStorage.x function helpers that JSON.parse/stringify for us

Using this helper almost certainly simplifies the component in which I'm using it. I get to use what is basically just the `useState` API in my component. It handles state management AND persistence to localStorage, and reconciling the two. Having a simple helper would still require me to deal with both. It would almost certainly lead to a more complex component.

I will remove the listener to simplify it for the time being.

Reply:
3e903957fe3d49ec8275f1b212a1371aae4d0343

Reply:
3ad595fddeb127b8673d963fbe7039549a2104a8

Reply:
Ya that's a fair approach, apologies for making you remove the listener - totally not my intention, if you'd prefer to leave it in feel free to revert the commit, I should have made it clearer that I was thinking out loud / in a vacuum rather than making a hard change request 🙇‍♀️

Reply:
Nah, I kinda wanted to remove it when I pulled this code in originally. I kept it "just because". Hearing you point this stuff out made me think that we should just remove it. Less is more. Thanks for input!